libsql-sqlite3/src/vector.c

```diff
@@ -0,0 +1,558 @@
+    sqlite3_free(pzErrMsg);
+    return;
+  }
+  pVector = vectorContextAlloc(context, type, dims);
+  if( pVector==NULL ){
+    return;
+  }
+  if( vectorParse(argv[0], pVector, &pzErrMsg) != 0 ){
+    sqlite3_result_error(context, pzErrMsg, -1);
+    sqlite3_free(pzErrMsg);
+    goto out_free_vec;
+  }
+  vectorSerialize(context, pVector);
+out_free_vec:
+  vectorFree(pVector);
+}
+
+static void vector32Func(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  vectorFuncHintedType(context, argc, argv, VECTOR_TYPE_FLOAT32);
+}
+static void vector64Func(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  vectorFuncHintedType(context, argc, argv, VECTOR_TYPE_FLOAT64);
+}
+
+/*
+** Implementation of vector_extract(X) function.
+*/
+static void vectorExtractFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  char *pzErrMsg = 0;
+  Vector *pVector;
+  unsigned i;
+  int type, dims;
+
+  if( argc < 1 ){
+    return;
+  }
+  if( detectVectorParameters(argv[0], 0, &type, &dims, &pzErrMsg) != 0 ){
+    sqlite3_result_error(context, pzErrMsg, -1);
+    sqlite3_free(pzErrMsg);
+    return;
+  }
+  pVector = vectorContextAlloc(context, type, dims);
+  if( pVector==NULL ){
+    return;
+  }
+  if( vectorParse(argv[0], pVector, &pzErrMsg)<0 ){
+    sqlite3_result_error(context, pzErrMsg, -1);
+    sqlite3_free(pzErrMsg);
+    goto out_free;
+  }
+  vectorMarshalToText(context, pVector);
+out_free:
+  vectorFree(pVector);
+}
+
+/*
+** Implementation of vector_distance_cos(X, Y) function.
+*/
+static void vectorDistanceCosFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  char *pzErrMsg = 0;
+  Vector *pVector1, *pVector2;
+  int type1, type2;
+  int dims1, dims2;
+  if( argc < 2 ) {
+    return;
+  }
+  if( detectVectorParameters(argv[0], 0, &type1, &dims1, &pzErrMsg) != 0 ){
+    sqlite3_result_error(context, pzErrMsg, -1);
+    sqlite3_free(pzErrMsg);
+    goto out_free;
+  }
+  if( detectVectorParameters(argv[1], 0, &type2, &dims2, &pzErrMsg) != 0 ){
+    sqlite3_result_error(context, pzErrMsg, -1);
+    sqlite3_free(pzErrMsg);
+    goto out_free;
+  }
+  pVector1 = vectorContextAlloc(context, type1, dims1);
+  if( pVector1==NULL ){
+    goto out_free;
+  }
+  pVector2 = vectorContextAlloc(context, type2, dims2);
+  if( pVector2==NULL ){
+    goto out_free;
+  }
+  if( vectorParse(argv[0], pVector1, &pzErrMsg)<0 ){
+    sqlite3_result_error(context, pzErrMsg, -1);
+    sqlite3_free(pzErrMsg);
+    goto out_free;
+  }
+  if( vectorParse(argv[1], pVector2, &pzErrMsg)<0 ){
+    sqlite3_result_error(context, pzErrMsg, -1);
+    sqlite3_free(pzErrMsg);
+    goto out_free;
+  }
+  if( dims1 != dims2 ){
+    sqlite3_result_error(context, "vectors must have the same length", -1);
+    goto out_free;
+  }
+  if( type1 != type2 ){
+    sqlite3_result_error(context, "vectors must have the same type", -1);
+    goto out_free;
+  }
```

Comment:
Would it make sense to check this before parsing vectors?

Reply:
Yes, sure, we can make this check earlier